services/venice-controller/src/main/java/com/linkedin/venice/controller/stats/SparkServerStats.java

```diff
@@ -1,57 +1,186 @@
+        otelRepository,
+        this::registerSensor,
+        ControllerTehutiMetricNameEnum.CURRENT_IN_FLIGHT_REQUEST,
+        Collections.singletonList(new OccurrenceRate()),
+        baseDimensionsMap,
+        ControllerRoute.class);
+
+    successfulRequestCountMetric = MetricEntityStateFourEnums.create(
+        ControllerMetricEntity.CALL_COUNT.getMetricEntity(),
+        otelRepository,
+        this::registerSensor,
+        ControllerTehutiMetricNameEnum.SUCCESSFUL_REQUEST,
+        Collections.singletonList(new OccurrenceRate()),
+        baseDimensionsMap,
+        ControllerRoute.class,
+        HttpResponseStatusEnum.class,
+        HttpResponseStatusCodeCategory.class,
+        VeniceResponseStatusCategory.class);
+
+    failedRequestCountMetric = MetricEntityStateFourEnums.create(
+        ControllerMetricEntity.CALL_COUNT.getMetricEntity(),
+        otelRepository,
+        this::registerSensor,
+        ControllerTehutiMetricNameEnum.FAILED_REQUEST,
+        Collections.singletonList(new OccurrenceRate()),
+        baseDimensionsMap,
+        ControllerRoute.class,
+        HttpResponseStatusEnum.class,
+        HttpResponseStatusCodeCategory.class,
+        VeniceResponseStatusCategory.class);
+
+    successfulRequestLatencyHistogramMetric = MetricEntityStateFourEnums.create(
+        ControllerMetricEntity.CALL_TIME.getMetricEntity(),
+        otelRepository,
+        this::registerSensor,
+        ControllerTehutiMetricNameEnum.SUCCESSFUL_REQUEST_LATENCY,
+        Collections.singletonList(new OccurrenceRate()),
+        baseDimensionsMap,
+        ControllerRoute.class,
+        HttpResponseStatusEnum.class,
+        HttpResponseStatusCodeCategory.class,
+        VeniceResponseStatusCategory.class);
+
+    failedRequestLatencyHistogramMetric = MetricEntityStateFourEnums.create(
+        ControllerMetricEntity.CALL_TIME.getMetricEntity(),
+        otelRepository,
+        this::registerSensor,
+        ControllerTehutiMetricNameEnum.FAILED_REQUEST_LATENCY,
+        Collections.singletonList(new OccurrenceRate()),
+        baseDimensionsMap,
+        ControllerRoute.class,
+        HttpResponseStatusEnum.class,
+        HttpResponseStatusCodeCategory.class,
+        VeniceResponseStatusCategory.class);
+
   }
 
-  public void recordRequest() {
-    requests.record();
-    currentInFlightRequestTotal.record(1);
+  public void recordRequest(Request request) {
+    ControllerRoute route = Objects.requireNonNull(ControllerRoute.valueOfPath(request.uri()));
+    requests.record(1);
+    inFlightRequestTotalCountMetric.record(1, route);
   }
 
-  public void recordSuccessfulRequestLatency(double latency) {
-    finishRequest();
-    successfulRequest.record();
-    successfulRequestLatency.record(latency);
+  public void recordSuccessfulRequest(Request request, Response response, double latency) {
+    ControllerRoute route = Objects.requireNonNull(ControllerRoute.valueOfPath(request.uri()));
```

Comment:
can `ControllerRoute.valueOfPath(request.uri())` be null? If the uri is not from one of the defined ones? We should not crash here in that case. In that case, we should introduce a fall back like `UNKNOWN_API` instead.